src/EditorFeatures/Core/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs

```diff
@@ -101,33 +104,35 @@ private bool TryStart(CancellationToken cancellationToken)
 
                 OpeningPoint = SubjectBuffer.CurrentSnapshot.CreateTrackingPoint(openingSnapshotPoint, PointTrackingMode.Positive);
 
-                var context = GetBraceCompletionContext();
-                if (context == null)
+                var document = SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                if (document == null)
                 {
                     return false;
                 }
 
-                var braceResult = _service.GetBraceCompletionAsync(context.Value, cancellationToken).WaitAndGetResult(cancellationToken);
-                if (braceResult == null)
+                var parsedDocument = ParsedDocument.CreateSynchronously(document, cancellationToken);
```

Comment:
aside: i wonder if internally we should CWT from Doc to ParsedDoc.  it seems likely that if 1 command handler is getting this, others will as well, and this will ensure they share all that computation and there's no chance of one gc'ing and then recomputing.

Reply:
Presumably only a single command handler is running at any given time since they all run on UI thread. Also likely that these commands actually make a change to the text buffer and thus next handler needs a new document snapshot anyways. Transition from one document snapshot to the next should use incremental parsing though.